xds/src/main/java/io/grpc/xds/Bootstrapper.java

```diff
@@ -125,9 +120,163 @@ private static Bootstrap readConfig() throws IOException {
+      Struct.Builder structBuilder = Struct.newBuilder();
+      for (Map.Entry<String, ?> entry : metadata.entrySet()) {
+        structBuilder.putFields(entry.getKey(), convertToValue(entry.getValue()));
+      }
+      nodeBuilder.setMetadata(structBuilder);
+    }
+    Map<String, ?> rawLocality = (Map<String, ?>) rawNode.get("locality");
+    if (rawLocality != null) {
+      Locality.Builder localityBuilder = Locality.newBuilder();
+      String region = (String) rawLocality.get("region");
+      if (region == null) {
+        throw new IOException("Invalid bootstrap: malformed 'node : locality'.");
+      }
+      localityBuilder.setRegion((String) rawLocality.get("region"));
+      if (rawLocality.containsKey("zone")) {
+        localityBuilder.setZone((String) rawLocality.get("zone"));
+      }
+      if (rawLocality.containsKey("sub_zone")) {
+        localityBuilder.setSubZone((String) rawLocality.get("sub_zone"));
+      }
+      nodeBuilder.setLocality(localityBuilder);
+    }
+    String buildVersion = (String) rawNode.get("build_version");
+    if (buildVersion != null) {
+      nodeBuilder.setBuildVersion(buildVersion);
+    }
+
+    return new BootstrapInfo(serverConfig, nodeBuilder.build());
+  }
+
+  /**
+   * Converts Java representation of the given JSON value to protobuf's {@link
+   * com.google.protobuf.Value} representation.
+   *
+   * <p>The given {@code rawObject} must be a valid JSON value in Java representation, which is
+   * either a {@code Map<String, ?>}, {@code List<?>}, {@code String}, {@code Double},
+   * {@code Boolean}, or {@code null}.
+   */
+  private static Value convertToValue(Object rawObject) {
+    Value.Builder valueBuilder = Value.newBuilder();
+    if (rawObject == null) {
+      valueBuilder.setNullValue(NullValue.NULL_VALUE);
+    } else if (rawObject instanceof Double) {
+      valueBuilder.setNumberValue((Double) rawObject);
+    } else if (rawObject instanceof String) {
+      valueBuilder.setStringValue((String) rawObject);
+    } else if (rawObject instanceof Boolean) {
+      valueBuilder.setBoolValue((Boolean) rawObject);
+    } else if (rawObject instanceof Map) {
+      Struct.Builder structBuilder = Struct.newBuilder();
+      @SuppressWarnings("unchecked")
+      Map<String, ?> map = (Map<String, ?>) rawObject;
+      for (Map.Entry<String, ?> entry : map.entrySet()) {
+        structBuilder.putFields(entry.getKey(), convertToValue(entry.getValue()));
+      }
+      valueBuilder.setStructValue(structBuilder);
+    } else if (rawObject instanceof List) {
+      ListValue.Builder listBuilder = ListValue.newBuilder();
+      List<?> list = (List<?>) rawObject;
+      for (Object obj : list) {
+        listBuilder.addValues(convertToValue(obj));
+      }
+      valueBuilder.setListValue(listBuilder);
+    }
+    return valueBuilder.build();
+  }
+
+  // TODO(chengyuanzhang): May need more complex structure for channel creds config representation.
+  static class ChannelCreds {
```

Comment:
Singular form ChannelCred?

Reply:
Isn't creds for credentials always used in plural as credentials stand for a set of attestations (e.g., authority, authenticity, etc)?